### ISSUE TITLE
Wrong registry address with tfs npm

### ISSUE THREAD

**Issue by amiram** [pnpm/pnpm#718]
### pnpm version:
0.65.2

### Code to reproduce the issue:
Unknown

### Expected behavior:
Packages should install.

### Actual behavior:
Some of the packages could not be installed, sometimes all of them.

### Additional information:
We use local tfs npm registry. its address is written correctly in npm config:
registry = "http://ourhost/tfs/OurCollection/_packaging/npm-registry-name/npm/registry"
npm install is working fine.
When using pnpm, it is using the above url without the word "registry" at the end, so it gets 404 from tfs. The weird thing is that sometimes it is using the "registry" word for some of the packages, but it fails as soon as it is not using the word for a certain package.
There were some attempts that I could say that it was using the right address for packages starting with a "@" and the wrong one for those that don't start with it, but it is not consistent.

 - `node -v` prints:
v6.9.5
 - Windows, OS X, or Linux?:
Bash on Ubuntu on Windows

**Comment by zkochan** [comment]
I think I know the reason.

Will it work if you put a slash to the end of your registry URL?

`registry = "http://ourhost/tfs/OurCollection/_packaging/npm-registry-name/npm/registry/"`

**Comment by amiram** [comment]
It worked! I had to change the local .npmrc in my project (who put it there? is it pnpm?). Now I have 401 for all package downloads.

**Comment by zkochan** [comment]
no, pnpm does not create `.npmrc` files.

Regarding the 401. In what format do you specify authorization tokens?

**Comment by amiram** [comment]
[server_address]:_authToken=[very_long_token]

**Comment by zkochan** [comment]
Have you added the slash there as well?

**Comment by amiram** [comment]
I was there from the beginning, however it was without the "registry" word. I added it and it is also not working.

**Comment by zkochan** [comment]
We use [this package](https://github.com/rexxars/registry-auth-token) for getting the tokens from configs. I assume it cannot find the token for some reason

**Comment by zkochan** [comment]
> [server_address]:_authToken=[very_long_token]

Do you specify the protocol here or just start with `//`

Seems like it works when the protocol is relative but fails in other case.

**Comment by amiram** [comment]
Starts with `//`

**Comment by zkochan** [comment]
Could you then post exactly how the `[server_address]:_authToken=` part looks like and the `registry = [registry url]`

Because I cannot reproduce with `//`

**Comment by amiram** [comment]
Just found out that this package gives you the right token for requests made to get packages descriptions but not for download. The reason is that the token is for such a url:
http://host/tfs/MyCollection/_packaging/MyNpmRegistry/npm/ (which is good also if you add the "registry")
but in tfs npm packages are downloaded from:
http://host/tfs/MyCollection/_packaging/[pkgid]/npm/registry/[pkgname]

so I shortened the token url to end with "_packaging" and it worked! However, tfs instructs you to use the token with the full url, so maybe we need to save the token when we get the package description and use it to download,

**Comment by zkochan** [comment]
Good investigation!

Seems like npm solves it by using the [always-auth](https://docs.npmjs.com/cli/adduser#always-auth) config

> This will ensure that all requests to that registry (including for tarballs) include an authorization header. This setting may be necessary for use with private registries where metadata and package tarballs are stored on hosts with different hostnames.

We can probably do the same

**Comment by amiram** [comment]
After little investigation, we can do it if we put the auth data in PackageMeta, then let the resolve module put it in the ResolveResult so we can pass it to the fetchToStore function. From there we can pass it to the Got module so it won't try to find the auth data by itself but use the given one. What do you think?

**Comment by zkochan** [comment]
Sounds right, if there cannot be situation where the 2nd endpoint should have a different auth token. We can check the auth token of the 2nd endpoint first and if there is not token for it, just use the 1st endpoint's auth token.

Also we need to be sure that this tokens are not logged in `pnpm-debug.log`

**Comment by zkochan** [comment]
Partially fixed. When the tarball path is in `shrinkwrap.yaml` alread and there is no need for a request for the package meta information. It doesn't work

**Comment by zkochan** [comment]
Excuse me for the redundant noise. This is fixed now. Added additional tests to confirm all cases

**Comment by amiram** [comment]
wow! so fast! great work!
BTW, I tested pnpm on the pnpm code itself. My os is Windows :(. However, I use bash on Ubuntu on Windows as default shell. Looks like it doesn't save space at all. node_modules is 104MB. With regular npm it is 105MB. I see many unrecognized links inside (probably windows can't read them since ubuntu creates them). Most of the space is taken by the folder .registry.npmjs.org (even that I use the private tfs npm).
There is a folder just like that in the npm store in my pc (it is a copy). In another angular project which has some public and private packages, the folder in node_modules has our npm address, but still it doesn't save space.

**Comment by zkochan** [comment]
Awesome!

Regarding disk space. I don't know how it works via Ubuntu on Windows, I assume the same way just the store is in a different location, not where on Windows. However, the 100MB space that you see is used in the project's node_modules is the same 100MB that you see is used by the store. This is how hard links look like. It seems they use space in several places on the disk but indeed they use the same disk space. If you look at the total disk space though, it should count 100MB only once, the 100MB that is in the store. More details about this: [Why do hard links seem to take the same space as the originals?](https://unix.stackexchange.com/questions/88423/why-do-hard-links-seem-to-take-the-same-space-as-the-originals)

**Comment by amiram** [comment]
I have a special drive for code and the project is beside the store. Both can be accessed by windows and ubuntu. I'm sure it wasn't a hard link. I deleted it from the store and it was still present in node_modules

**Comment by zkochan** [comment]
A file is removed only when all hard links pointing to it are removed

And it is impossible that not hard links are created. If hard link cannot be created, pnpm fails. See issue #712

**Comment by amiram** [comment]
Oh good to know. BTW, there are files in node_modules that Windows can't read. Is the link created by ubuntu is cross-platform? If not, how windows can use the hard link to the store?

**Comment by zkochan** [comment]
For this I currently don't know the answer. I personally had issues with Ubuntu on Windows even when using npm, so I don't use it. You can create a separate issue for this

**Comment by NCC1701M** [comment]
Still doesn't work with Version 1.12.0
My $home/.npmrc files looks like 
```
registry=http://myhost:8080/tfs/Projects/_packaging/my-packages/npm/registry
always-auth=true
//myhost:8080/tfs/Projects/_packaging/my-packages/npm/:_authToken=[mytoken]
```

When I execute `npm i my-package` I get the following log file:
```
---
  0 debug pnpm:progress: 
    status: "installing"
    pkg: 
      rawSpec: "^1.1.2"
      name: "my-package"
  1 debug pnpm:progress: 
    status: "resolving"
    pkg: 
      rawSpec: "^1.1.2"
      name: "my-package"
  2 debug pnpm:registry: "http request uri http://myhost:8080/tfs/Projects/_packaging/my-packages/npm/registry/my-package"
  3 debug pnpm:registry: "http request always-auth set; sending authorization"
  4 info pnpm:registry: "attempt registry request try #1 at 11:18:57"
  5 info pnpm:registry: "retry will retry, error on last attempt: Error: This request requires auth credentials. Run `npm login` and repeat the request."
  6 info pnpm:registry: "attempt registry request try #2 at 11:19:07"
  7 info pnpm:registry: "retry will retry, error on last attempt: Error: This request requires auth credentials. Run `npm login` and repeat the request."
  8 info pnpm:registry: "attempt registry request try #3 at 11:20:07"
  9 debug pnpm:progress: 
    status: "error"
    pkg: 
      rawSpec: "^1.1.2"
      name: "my-package"
  10 error pnpm: 
    message: {}
    err: 
      name: "Error"
      message: "This request requires auth credentials. Run `npm login` and repeat the request."
      stack: "Error: This request requires auth credentials. Run `npm login` and repeat the request.\n
		    at RegClient.authify (C:\\Users\\user123\\AppData\\Roaming\\npm\\node_modules\\pnpm\\lib\\node_modules\\npm-registry-client\\lib\\authify.js:17:14)\n
		    at RegClient.makeRequest (C:\\Users\\user123\\AppData\\Roaming\\npm\\node_modules\\pnpm\\lib\\node_modules\\npm-registry-client\\lib\\request.js:116:17)\n
		    at C:\\Users\\user123\\AppData\\Roaming\\npm\\node_modules\\pnpm\\lib\\node_modules\\npm-registry-client\\lib\\request.js:68:17\n
		    at RetryOperation._fn (C:\\Users\\user123\\AppData\\Roaming\\npm\\node_modules\\pnpm\\lib\\node_modules\\npm-registry-client\\lib\\attempt.js:18:5)\n
		    at Timeout._onTimeout (C:\\Users\\user123\\AppData\\Roaming\\npm\\node_modules\\pnpm\\lib\\node_modules\\retry\\lib\\retry_operation.js:68:10)\n
		    at ontimeout (timers.js:469:11)\n
		    at tryOnTimeout (timers.js:304:5)\n
		    at Timer.listOnTimeout (timers.js:264:5)"

```
